<a href="https://colab.research.google.com/github/mlacasa/1BATXILLERAT/blob/main/Pr%C3%A0cticaBasedeDades.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PRÀCTICA ASSIGNATURA MATEMÀTIQUES APLICADA 1 BAT
## Marcos Lacasa

1- Carreguem l'arxiu "Distrubución Salarios" de open-acces de l'agència tributària.



In [1]:
# Importem llibreria pandas
import pandas as pd

In [2]:
# definim el path des del nostre drive i carreguem el format csv.
path = '/content/drive/MyDrive/Colab Notebooks/BAT/data2.csv'
data = pd.read_csv(path)

In [3]:
data.head()

ID         2001         2002         2003         2004         2005  \
0   1   29,541,615   27,645,802   29,437,929   27,855,280   31,801,465   
1   2   72,965,628   81,467,945   87,340,372   68,217,591   71,854,670   
2   3  118,397,894  117,964,368  113,769,342  102,460,947  109,068,935   
3   4  149,288,420  145,505,890  153,889,058  144,327,889  154,811,651   
4   5  187,829,221  179,214,060  177,103,986  164,526,172  179,556,090   

          2006         2007         2008         2009  ... 2013.4 2014.4  \
0   31,584,347   25,499,488   27,306,225   28,879,391  ...     93     94   
1   72,813,778   59,420,277   62,227,592   67,221,958  ...    296    295   
2  108,110,010   88,723,845   89,901,678   99,150,785  ...    498    497   
3  154,349,596  130,103,812  129,261,037  140,721,048  ...    702    701   
4  174,609,190  143,789,890  140,636,251  159,501,515  ...    896    898   

  2015.4 2016.4 2017.4 2018.4 2019.4 2020.4 2021.4 2022.4  
0     95     96     97     98     98    100     99     98  
1    296    296    296    296    295    296    297    292  
2    498    500    498    499    498    499    498    497  
3    700    699    697    697    700    700    699    699  
4    897    895    897    896    898    900    899    899  

[5 rows x 111 columns]

In [4]:
data.shape

(400, 111)

### OBSERVACIONS

1- Comprobem que les dades son 'objectes' i els nombres tenen 2 commes.
2- Hem de suprimir les commes de totes les columnes.

Hi ha 400 files (una per cada tram) i 111 columnes (que son 5 categoríes per cada 22 anys analitzats.

In [5]:
# Convertim totes les columnes en "string" per poder utilitzar el mètode "replace"
data = data.astype(str)

In [6]:
# Treiem les commes però ho fem amb un loop utilitzant 'iloc' de pandas
for i in range(1,111):
  data.iloc[:,i]=data.iloc[:,i].str.replace(',','')

### documentació iloc de pandas

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

Visualitzar!!

In [7]:
# Seleccionem només les 3 primeres categoríes
# Les altres dues, son vectors LD, 22x3+1 = 67 colummes
data1 = data.iloc[:,0:67]

In [8]:
# Convertim totes les columnes com nombres naturals (enters)
data1 = data1.astype(int)

In [9]:
# comprobem i visualitzem les 5 primers registres
data1.head()

ID       2001       2002       2003       2004       2005       2006  \
0   1   29541615   27645802   29437929   27855280   31801465   31584347   
1   2   72965628   81467945   87340372   68217591   71854670   72813778   
2   3  118397894  117964368  113769342  102460947  109068935  108110010   
3   4  149288420  145505890  153889058  144327889  154811651  154349596   
4   5  187829221  179214060  177103986  164526172  179556090  174609190   

        2007       2008       2009  ...  2013.2  2014.2  2015.2  2016.2  \
0   25499488   27306225   28879391  ...  384262  369837  360424  359341   
1   59420277   62227592   67221958  ...  267525  263879  269745  270013   
2   88723845   89901678   99150785  ...  227742  230155  235644  242355   
3  130103812  129261037  140721048  ...  219984  217609  228211  242164   
4  143789890  140636251  159501515  ...  204015  202091  214043  214446   

   2017.2  2018.2  2019.2  2020.2  2021.2  2022.2  
0  336325  328778  310946  270572  276001  244376  
1  248893  233521  210956  194006  200752  186871  
2  226109  210252  180277  170422  164307  155482  
3  208225  192375  172495  164014  157221  149360  
4  191359  186405  164155  164394  146657  139655  

[5 rows x 67 columns]

## Construcció de vectors

Volem construir v4 vectors, amb una dimensió de 22 * 400 * 3 = 26.400 registres.
Però pot ser útil construir-los per parts.

Exemple

Diferències en l'ús dels mètodes 'append' i 'extend'

volem construir un vector que sigui:



```
vectorFinal = [1,2,3,1,2,3,1,2,3]
```

Observar que es repeteixen 3 cops els valors '1,2,3'

In [10]:
# Mètode 'append'
# Creem un vector inicial

v1 = [1,2,3]
# creem el vector final
vectorFinal = [] # Un vector buit
# utilitzem el mètode 'append'

vectorFinal.append([v1]*3)

In [11]:
# No és exactament el que volem.
vectorFinal

[[[1, 2, 3], [1, 2, 3], [1, 2, 3]]]

In [12]:
#Hem creat un vector composat per altre vector que conté 3 vectors.
len(vectorFinal)

1

In [13]:
len(vectorFinal[0])

3

In [14]:
vectorFinal[0][0] # Observar que hi ha 2 nivells.

[1, 2, 3]

In [15]:
# Mètode 'extend'
v1 = [1,2,3]
# creem el vector final
vectorFinal = [] # Un vector buit
# utilitzem el mètode 'append'

vectorFinal.extend(v1)
vectorFinal.extend(v1)
vectorFinal.extend(v1)

In [16]:
# Ara sí que és el que volem.
vectorFinal

[1, 2, 3, 1, 2, 3, 1, 2, 3]

In [17]:
v1 = [1,2,3] # creo dos vectors diferents
v2 = [4,5,6]

v3 = []
v3.extend(v1) # afegir els vectos a altre general v3
v3.extend(v2)
v3 # Observar el resultat, un sol vector.


[1, 2, 3, 4, 5, 6]

In [18]:
v1 = [1,2,3]
v2 = [4,5,6]

v3 = []
v3.append(v1) # ara utilitzo el mètode append
v3.append(v2)
v3            # El resultat és 2 vectors.

[[1, 2, 3], [4, 5, 6]]

## Convertir un vector de vectors en un de sol

Ara hem vist la diferència entre els mètodes 'append' i 'extend'.\
Però ens pot ser molt útil crear un vector de vectors, mitjançant un 'for'.\

### Mètode 'flatten'

Aquest mètode pertany a una llibreria específica per treballar amb vectors, que és numpy. Anem a estudiar-ho.

In [19]:
import numpy as np    # Importem la llibreria numpy

myarray = np.array([[1,2,3],[4,5,6]])  # creem un vector de vectors, i array = vector

# Convertim a un sol vector
myarray.flatten('A')

array([1, 2, 3, 4, 5, 6])

Els paràmetres de flatten són:

"C" significa aplanar en ordre de fila principal (estil C). \
"F" significa aplanar en ordre de columna major (estil Fortran). \
"A" significa aplanar en l'ordre principal de la columna si a és Fortran contigu a la memòria, en cas contrari. \
"K" significa aplanar a en l'ordre en què apareixen els elements a la memòria.

El valor predeterminat és "C".

In [20]:
myarray.flatten('C') # Escollim C per defecte

array([1, 2, 3, 4, 5, 6])

In [21]:
myarray.flatten() # Escollim C per defecte

array([1, 2, 3, 4, 5, 6])

In [22]:
# Recordem
v3

[[1, 2, 3], [4, 5, 6]]

In [23]:
# Treballem amb numpy, convertim a array i utilitzem el mètode flatten
np.array(v3).flatten()

array([1, 2, 3, 4, 5, 6])

In [24]:
# Així podem fer:
vF = np.array(v3).flatten()
vF

array([1, 2, 3, 4, 5, 6])

In [25]:
data1.dtypes # Avaluem el tipus de cada columna

ID        int64
2001      int64
2002      int64
2003      int64
2004      int64
          ...  
2018.2    int64
2019.2    int64
2020.2    int64
2021.2    int64
2022.2    int64
Length: 67, dtype: object

## Preparem el format per analitzar i visualitzar

Hem de crear la dataframe amb el disseny que ens interessa

In [26]:
# Vector amb els anys, del 2001-2022
years = list(range(2001,2023))
print("el primer any és: ",years[0]," i l'últim és ", years[21])

el primer any és:  2001  i l'últim és  2022


In [27]:
# Els noms de les columnes de la nostre DataFrame
nomsCol = ['rank', 'value', 'year', 'category']

In [28]:
data2 = pd.DataFrame(columns = nomsCol)

In [29]:
data2

Empty DataFrame
Columns: [rank, value, year, category]
Index: []

In [30]:
# Les 3 categories son:
cat = ['Salary', 'tax', 'employees']

In [31]:
v4=[[cat[0]]*10]    # deu cops la paraula 'Salary'
v5 = [[cat[1]]*10]
v6 = [[cat[2]]*10]

In [32]:
v4 # Ull viu, es un vector amb altre vector.

[['Salary',
  'Salary',
  'Salary',
  'Salary',
  'Salary',
  'Salary',
  'Salary',
  'Salary',
  'Salary',
  'Salary']]

In [33]:
v7=[]
v7.extend(v4)
v7.extend(v5)
v7.extend(v6)
len(v7)

3

In [34]:
v7[2] # Una mica llarg

['employees',
 'employees',
 'employees',
 'employees',
 'employees',
 'employees',
 'employees',
 'employees',
 'employees',
 'employees']

In [35]:
np.array(cat) # Convertim a array mitjançant np.array()

array(['Salary', 'tax', 'employees'], dtype='<U9')

In [36]:
# Volem crear un vector que repeteixi 10 cops les categories, que tingui una llargada de 30 en un sol vector.
[[x]*2 for x in cat]  # utilitzem la "list comprehension"

# Llegir https://ellibrodepython.com/list-comprehension-python


[['Salary', 'Salary'], ['tax', 'tax'], ['employees', 'employees']]

In [37]:
#Convertim en un sol vector en una sola línea de codi
np.array([[x]*10 for x in cat]).flatten()

array(['Salary', 'Salary', 'Salary', 'Salary', 'Salary', 'Salary',
       'Salary', 'Salary', 'Salary', 'Salary', 'tax', 'tax', 'tax', 'tax',
       'tax', 'tax', 'tax', 'tax', 'tax', 'tax', 'employees', 'employees',
       'employees', 'employees', 'employees', 'employees', 'employees',
       'employees', 'employees', 'employees'], dtype='<U9')

## Exercici

1- Configurar la 'data2' de la següent manera:

- Ha de tenir tots els valors de 'data1' disposats en 4 columnes.
- Heu de dissenyar una funció per tal de recòrrer les columnes i els valors.
- cada valor ha d'estar definit per un rank, value, year i category.
- Les 3 categoríes son al vector 'cat'.
- Els anys són al vector 'years'.

Exemple del 1er registre:

rank | value | year | category  
--- | --- |--- |--- |
1|29541615|2001|Salary


### Notes

- Les columnes amb els anys pertànyen:\
1- 2001 -> Salary\
2- 2001.1 -> tax\
3- 2001.2 -> employees

### Indicacions

Realitzar els loops amb diferents "for" per recòrrer les columnes. Utilitzar adequadament el mètode 'iloc' de pandas.

Recordar:


```
df.iloc[1:3, 0:3]
```
vol dir, escull les files 1,2 i les columnes 0,1,2.


```
df.iloc[:, 1]
```
vol dir, escull totes les files de la columna 2.


In [38]:
# El primer valor del vector years és:
years[0]

2001

In [39]:
# Un vector que 400 registres de cada any:
len(years) # Mida del vector

22

In [40]:
AllYF = []                    # Creo el vector amb 22 vectors amb 400 per cada any
for i in range(len(years)):   # Recorrer els 22 anys
  AllYF.append([years[i]]*400)# Afegeix un vector de cada any amb 400 repeticions
# Ordre: append - afegir vector a AllYF
# [years[i]]*400, fes un vector amb 400 cops el valor years[i]

In [41]:
print("El vector medeix (té nombre de vectors): ", len(AllYF), " i cada vector té una mida de ",len(AllYF[0]))

El vector medeix (té nombre de vectors):  22  i cada vector té una mida de  400


### Calcular el nombre de files

400 registres de cada any i categoria\
22 anys\
3 categories\
Total = 400·22·3 = 26400 registres


In [42]:
# Càlcul del vector Categories
import numpy as np
data2 = pd.DataFrame(columns = nomsCol)

In [43]:
data2['category'] = np.array([[x]*8800 for x in cat]).flatten()
data2['rank'] = np.array([np.arange(1,401,1,dtype=int)]*66).flatten()
data2['year'] = np.array([np.array([[x]*400 for x in np.array([np.arange(2001,2023,1,dtype=int)]).flatten()]).flatten()]*3).flatten()

In [44]:
data2.shape

(26400, 4)

In [45]:
data2.iloc[10000,:]

rank           1
value        NaN
year        2004
category     tax
Name: 10000, dtype: object

### Columna Valors

In [46]:
data1.columns

Index(['ID', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
       '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
       '2018', '2019', '2020', '2021', '2022', '2001.1', '2002.1', '2003.1',
       '2004.1', '2005.1', '2006.1', '2007.1', '2008.1', '2009.1', '2010.1',
       '2011.1', '2012.1', '2013.1', '2014.1', '2015.1', '2016.1', '2017.1',
       '2018.1', '2019.1', '2020.1', '2021.1', '2022.1', '2001.2', '2002.2',
       '2003.2', '2004.2', '2005.2', '2006.2', '2007.2', '2008.2', '2009.2',
       '2010.2', '2011.2', '2012.2', '2013.2', '2014.2', '2015.2', '2016.2',
       '2017.2', '2018.2', '2019.2', '2020.2', '2021.2', '2022.2'],
      dtype='object')

In [47]:
data1.head()

ID       2001       2002       2003       2004       2005       2006  \
0   1   29541615   27645802   29437929   27855280   31801465   31584347   
1   2   72965628   81467945   87340372   68217591   71854670   72813778   
2   3  118397894  117964368  113769342  102460947  109068935  108110010   
3   4  149288420  145505890  153889058  144327889  154811651  154349596   
4   5  187829221  179214060  177103986  164526172  179556090  174609190   

        2007       2008       2009  ...  2013.2  2014.2  2015.2  2016.2  \
0   25499488   27306225   28879391  ...  384262  369837  360424  359341   
1   59420277   62227592   67221958  ...  267525  263879  269745  270013   
2   88723845   89901678   99150785  ...  227742  230155  235644  242355   
3  130103812  129261037  140721048  ...  219984  217609  228211  242164   
4  143789890  140636251  159501515  ...  204015  202091  214043  214446   

   2017.2  2018.2  2019.2  2020.2  2021.2  2022.2  
0  336325  328778  310946  270572  276001  244376  
1  248893  233521  210956  194006  200752  186871  
2  226109  210252  180277  170422  164307  155482  
3  208225  192375  172495  164014  157221  149360  
4  191359  186405  164155  164394  146657  139655  

[5 rows x 67 columns]

In [48]:
# Tenim 66 columnas que contenen els valors demanats.
# Una columna es pot cridar com
vValues=[]
for i in range(1,67):
  vValues.append(np.array(data1.iloc[:,i]))

In [49]:
len(np.array(vValues).flatten())

26400

In [50]:
data2['value'] = np.array(vValues).flatten()

In [51]:
data2

rank      value  year   category
0         1   29541615  2001     Salary
1         2   72965628  2001     Salary
2         3  118397894  2001     Salary
3         4  149288420  2001     Salary
4         5  187829221  2001     Salary
...     ...        ...   ...        ...
26395   396       2959  2022  employees
26396   397       3151  2022  employees
26397   398       3090  2022  employees
26398   399       2951  2022  employees
26399   400       3013  2022  employees

[26400 rows x 4 columns]

In [52]:
import plotly.express as px

# Realitzar gràfic animat amb valors i anys

1- Estudia la llibreria Plotly\

Omple els arguments:



```
(df, x="continent", y="pop", color="continent",
  animation_frame="year", animation_group="country", range_y=[0,4000000000])
```



x=\
y=\
color=\
animation_frame='year'\
range_y= Avaluar el rang

In [58]:
data2.describe()

rank         value          year
count  26400.00000  2.640000e+04  26400.000000
mean     200.50000  3.168356e+08   2011.500000
std      115.47188  9.204423e+08      6.344409
min        1.00000  5.060000e+02   2001.000000
25%      100.75000  8.569075e+04   2006.000000
50%      200.50000  8.273692e+07   2011.500000
75%      300.25000  2.649101e+08   2017.000000
max      400.00000  4.172739e+10   2022.000000

In [59]:
fig = px.bar(data2[data2['category']=='Salary'], x="rank", y="value", color="rank",
  animation_frame="year", animation_group="rank", range_y=[0,22e8])
fig.show()

In [60]:
fig = px.bar(data2[data2['category']=='employees'], x="rank", y="value", color="rank",
  animation_frame="year", animation_group="rank", range_y=[0,300000])
fig.show()

In [64]:
# NOs da problemas Orange3 con los años, hay que trabajarlos.
import datetime as td
pd.to_datetime('2018', format='%Y')

Timestamp('2018-01-01 00:00:00')

In [67]:
data2['year'] = pd.to_datetime(data2['year'], format='%Y')

In [68]:
data2.head()

rank      value       year category
0     1   29541615 2001-01-01   Salary
1     2   72965628 2001-01-01   Salary
2     3  118397894 2001-01-01   Salary
3     4  149288420 2001-01-01   Salary
4     5  187829221 2001-01-01   Salary

In [69]:
data2.to_csv('/content/drive/MyDrive/Colab Notebooks/BAT/salary.csv')